In [0]:
%tensorflow_version 2.x

import tensorflow as tf
import tensorflow_datasets as tfds
import tensorflow_hub as hub

###  Please ensure tensorflow 2.0 version loaded.

###  command to install latest tensorflow

!pip install tensorflow --upgrade

### Please restart the runtime

In [4]:
# list out the datasets
tfds.list_builders()

['abstract_reasoning',
 'aeslc',
 'aflw2k3d',
 'amazon_us_reviews',
 'bair_robot_pushing_small',
 'big_patent',
 'bigearthnet',
 'billsum',
 'binarized_mnist',
 'binary_alpha_digits',
 'c4',
 'caltech101',
 'caltech_birds2010',
 'caltech_birds2011',
 'cars196',
 'cassava',
 'cats_vs_dogs',
 'celeb_a',
 'celeb_a_hq',
 'chexpert',
 'cifar10',
 'cifar100',
 'cifar10_1',
 'cifar10_corrupted',
 'clevr',
 'cmaterdb',
 'cnn_dailymail',
 'coco',
 'coco2014',
 'coil100',
 'colorectal_histology',
 'colorectal_histology_large',
 'curated_breast_imaging_ddsm',
 'cycle_gan',
 'deep_weeds',
 'definite_pronoun_resolution',
 'diabetic_retinopathy_detection',
 'downsampled_imagenet',
 'dsprites',
 'dtd',
 'dummy_dataset_shared_generator',
 'dummy_mnist',
 'emnist',
 'eurosat',
 'fashion_mnist',
 'flores',
 'food101',
 'gap',
 'gigaword',
 'glue',
 'groove',
 'higgs',
 'horses_or_humans',
 'image_label_folder',
 'imagenet2012',
 'imagenet2012_corrupted',
 'imagenet_resized',
 'imdb_reviews',
 'iris',
 '

In [0]:
(train_data, validation_data), test_data = tfds.load(name='imdb_reviews',
         split = (tfds.Split.TRAIN.subsplit([6,4]), tfds.Split.TEST),
         as_supervised = True)

In [13]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))

train_examples_batch

<tf.Tensor: id=388, shape=(10,), dtype=string, numpy=
array([b"As a lifelong fan of Dickens, I have invariably been disappointed by adaptations of his novels.<br /><br />Although his works presented an extremely accurate re-telling of human life at every level in Victorian Britain, throughout them all was a pervasive thread of humour that could be both playful or sarcastic as the narrative dictated. In a way, he was a literary caricaturist and cartoonist. He could be serious and hilarious in the same sentence. He pricked pride, lampooned arrogance, celebrated modesty, and empathised with loneliness and poverty. It may be a clich\xc3\xa9, but he was a people's writer.<br /><br />And it is the comedy that is so often missing from his interpretations. At the time of writing, Oliver Twist is being dramatised in serial form on BBC television. All of the misery and cruelty is their, but non of the humour, irony, and savage lampoonery. The result is just a dark, dismal experience: the story p

In [14]:
train_labels_batch

<tf.Tensor: id=389, shape=(10,), dtype=int64, numpy=array([1, 1, 1, 1, 1, 1, 0, 1, 1, 0])>

In [0]:
import tensorflow_hub as hub

pre_trained_model = hub.load("https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1")

hub_layer = hub.KerasLayer(pre_trained_model, input_shape=[], dtype=tf.string, trainable=True)

In [16]:
# Here i'm printing 2 reviews 
train_examples_batch[:2] 

<tf.Tensor: id=559, shape=(2,), dtype=string, numpy=
array([b"As a lifelong fan of Dickens, I have invariably been disappointed by adaptations of his novels.<br /><br />Although his works presented an extremely accurate re-telling of human life at every level in Victorian Britain, throughout them all was a pervasive thread of humour that could be both playful or sarcastic as the narrative dictated. In a way, he was a literary caricaturist and cartoonist. He could be serious and hilarious in the same sentence. He pricked pride, lampooned arrogance, celebrated modesty, and empathised with loneliness and poverty. It may be a clich\xc3\xa9, but he was a people's writer.<br /><br />And it is the comedy that is so often missing from his interpretations. At the time of writing, Oliver Twist is being dramatised in serial form on BBC television. All of the misery and cruelty is their, but non of the humour, irony, and savage lampoonery. The result is just a dark, dismal experience: the story pe

In [17]:
# above 2 moview reviews text converted to 2 vectors
hub_layer(train_examples_batch[:2])

<tf.Tensor: id=573, shape=(2, 20), dtype=float32, numpy=
array([[ 3.9819887 , -4.4838037 ,  5.177359  , -2.3643482 , -3.2938678 ,
        -3.5364532 , -2.4786978 ,  2.5525482 ,  6.688532  , -2.3076782 ,
        -1.9807833 ,  1.1315885 , -3.0339816 , -0.7604128 , -5.743445  ,
         3.4242578 ,  4.790099  , -4.03061   , -5.992149  , -1.7297493 ],
       [ 3.4232912 , -4.230874  ,  4.1488533 , -0.29553518, -6.802391  ,
        -2.5163853 , -4.4002395 ,  1.905792  ,  4.7512794 , -0.40538004,
        -4.3401685 ,  1.0361497 ,  0.9744097 ,  0.71507156, -6.2657013 ,
         0.16533905,  4.560262  , -1.3106939 , -3.1121316 , -2.1338716 ]],
      dtype=float32)>

In [19]:
# create a neural network

model = tf.keras.Sequential()

model.add(hub_layer)  # add input layer
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1, activation='sigmoid'))

model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
keras_layer (KerasLayer)     (None, 20)                400020    
_________________________________________________________________
dense (Dense)                (None, 16)                336       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 17        
Total params: 400,373
Trainable params: 400,373
Non-trainable params: 0
_________________________________________________________________


In [0]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [21]:
model.fit(train_data.shuffle(10000).batch(512),
          epochs=20,
          validation_data = validation_data.batch(512),
          verbose = 1)

Epoch 1/20
30/30 [==============================] - 8s 269ms/step - loss: 0.6985 - accuracy: 0.5635 - val_loss: 0.0000e+00 - val_accuracy: 0.0000e+00
Epoch 2/20
30/30 [==============================] - 7s 240ms/step - loss: 0.6324 - accuracy: 0.6458 - val_loss: 0.6160 - val_accuracy: 0.6633
Epoch 3/20
30/30 [==============================] - 7s 219ms/step - loss: 0.5988 - accuracy: 0.6836 - val_loss: 0.5879 - val_accuracy: 0.6973
Epoch 4/20
30/30 [==============================] - 7s 218ms/step - loss: 0.5758 - accuracy: 0.7128 - val_loss: 0.5631 - val_accuracy: 0.7215
Epoch 5/20
30/30 [==============================] - 7s 220ms/step - loss: 0.5428 - accuracy: 0.7389 - val_loss: 0.5372 - val_accuracy: 0.7441
Epoch 6/20
30/30 [==============================] - 7s 239ms/step - loss: 0.5123 - accuracy: 0.7660 - val_loss: 0.5107 - val_accuracy: 0.7633
Epoch 7/20
30/30 [==============================] - 7s 223ms/step - loss: 0.4840 - accuracy: 0.7891 - val_loss: 0.4824 - val_accuracy: 0.781

In [22]:
model.predict(['This is the worst movie I have ever seen'])

array([[0.17902257]], dtype=float32)

0.17902257 it is near to 0, it is negetive review

In [24]:
model.predict(['An excellent movie that I enjoyed a lot'])

array([[0.9986124]], dtype=float32)

0.9986124 it is near to 1, positive review

In [26]:
# you can pass a list of reviews
model.predict(['This is the worst movie I have ever seen',
              'An excellent movie that I enjoyed a lot',
              'How can one make such a horrible movie? there is no story, acting, direction everything is poor'
              ])

array([[0.1790226 ],
       [0.9986124 ],
       [0.01820755]], dtype=float32)